core/src/main/java/org/infinispan/persistence/manager/TracedPersistenceManager.java

```diff
@@ -0,0 +1,153 @@
+      InfinispanSpan span = requestStart("approximateSize");
+      CompletionStage<Long> operation = super.approximateSize(predicate, segments);
+      operation.whenComplete(requestEnd(span));
+      return operation;
+   }
+
+   @Override
+   public CompletionStage<Long> size(Predicate<? super StoreConfiguration> predicate, IntSet segments) {
+      InfinispanSpan span = requestStart("size");
+      CompletionStage<Long> operation = super.size(predicate, segments);
+      operation.whenComplete(requestEnd(span));
+      return operation;
+   }
+
+   @Override
+   public CompletionStage<Void> writeToAllNonTxStores(MarshallableEntry marshalledEntry, int segment,
+                                                      Predicate<? super StoreConfiguration> predicate, long flags) {
+      InfinispanSpan span = requestStart("writeToAllNonTxStores");
+      CompletionStage<Void> operation = super.writeToAllNonTxStores(marshalledEntry, segment, predicate, flags);
+      operation.whenComplete(requestEnd(span));
+      return operation;
+   }
+
+   @Override
+   public CompletionStage<Void> commitAllTxStores(TxInvocationContext<AbstractCacheTransaction> txInvocationContext,
+                                                  Predicate<? super StoreConfiguration> predicate) {
+      InfinispanSpan span = requestStart("commitAllTxStores");
+      CompletionStage<Void> operation = super.commitAllTxStores(txInvocationContext, predicate);
+      operation.whenComplete(requestEnd(span));
+      return operation;
+   }
+
+   @Override
+   public CompletionStage<Void> rollbackAllTxStores(TxInvocationContext<AbstractCacheTransaction> txInvocationContext,
+                                                    Predicate<? super StoreConfiguration> predicate) {
+      InfinispanSpan span = requestStart("rollbackAllTxStores");
+      CompletionStage<Void> operation = super.rollbackAllTxStores(txInvocationContext, predicate);
+      operation.whenComplete(requestEnd(span));
+      return operation;
+   }
+
+   @Override
+   public <K, V> CompletionStage<Void> writeEntries(Iterable<MarshallableEntry<K, V>> iterable,
+                                                    Predicate<? super StoreConfiguration> predicate) {
+      InfinispanSpan span = requestStart("writeEntries");
+      CompletionStage<Void> operation = super.writeEntries(iterable, predicate);
+      operation.whenComplete(requestEnd(span));
+      return operation;
+   }
+
+   @Override
+   public CompletionStage<Long> performBatch(TxInvocationContext<AbstractCacheTransaction> ctx,
+                                             BiPredicate<? super WriteCommand, Object> commandKeyPredicate) {
+      InfinispanSpan span = requestStart("performBatch");
+      CompletionStage<Long> operation = super.performBatch(ctx, commandKeyPredicate);
+      operation.whenComplete(requestEnd(span));
+      return operation;
+   }
+
+   private InfinispanSpan requestStart(String operationName) {
```

Comment:
It looks to me like a great idea. Actually, I had thought about it, but I didn't do it since I thought it was required to acquire the lock in order to synchronize the decorator and the decorated method. 
Because I think that by the time we access the enabled variable in the decorated method this value could be different. 
In any case we don't need strong consistency on tracing so I think it is safe to use the volatile enabled variable.

Reply:
Sorry I didn't notice that enabled was effective immutable (immutable after the start of the service)